Dockerfile

```diff
@@ -3,22 +3,24 @@ FROM node:16
 
 LABEL author="Samuel Alev <samuel.alev@protonmail.com>"
 
-# Install Chrome
-RUN apt-get update \
-    && apt-get install -y wget gnupg \
-    && wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
-    && sh -c 'echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google.list' \
-    && apt-get update \
-    && apt-get install -y google-chrome-stable fonts-ipafont-gothic fonts-wqy-zenhei fonts-thai-tlwg fonts-kacst fonts-freefont-ttf libxss1 \
-      --no-install-recommends \
-    && rm -rf /var/lib/apt/lists/*
+WORKDIR /app
 
+COPY ["package.json", "package-lock.json", "./"]
```

Comment:
The changes made to this file have no effect on the final result.
The `package.json` and `package-lock.json` are copied with `COPY . .`, only the `.*` files won't be copied over but we anyway don't need them.
Moving the workdir also won't have any effect on the final outcome.

Reply:
> The changes made to this file have no effect on the final result.
> The `package.json` and `package-lock.json` are copied with `COPY . .`, only the `.*` files won't be copied over but we anyway don't need them.
> Moving the workdir also won't have any effect on the final outcome.

Hello @SamuelAlev 

There is a pretty huge upside to the copying `package.json` and `package-lock.json` first.

Of course, you are right, `COPY . .` does that anyway, but when you copy these 2 files first, these 2 files will be cached and the changes on other files won't remove the caching, It makes the build process way faster that what it is right now.
Not only these 2 files, but since these 2 files are used for listing dependencies, It also makes the process of installing dependencies use cache.

Reply:
There is a pretty huge upside to copying `package.json` and `package-lock.json` first.

Of course, you are right, `COPY . .` does the something, and the whole change doesn't affect the final result.

This is a common pattern in Dockerfiles (in all languages). The `npm install` step takes a long time, but you only need to run it when the package dependencies change.

So it's typical to see one step that just installs dependencies, and a second step that adds the actual application, because it makes rebuilding the container go faster.

You're right that this is essentially identical if you're building the image once; you get the same filesystem contents out at the end.

Say this happens while you're working on the package, though. You've changed some `src/*.ts` files but haven't changed the package.json. Now you re-run `docker build`.
Docker notices that the `package*.json` files haven't changed, so it uses the same image layer it built the first time without re-running anything, and it also skips the npm install step (because it assumes running the same command on the same input filesystem produces the same output filesystem).
So this makes the second build run faster.

It is also mentioned in the official Dockerizing Node.JS documentation
https://nodejs.org/en/docs/guides/nodejs-docker-webapp/#creating-a-dockerfile

---

Moving `WORKDIR` to the top is just for better readability, I can put it back if you want to!

Reply:
Gotcha! then let's move the `RUN npm ci` just under `COPY ["package.json", "package-lock.json", "./"]` so the cache doesn't burst with the `COPY . .`.